### PR TITLE
Block Log Tool

### PR DESCRIPTION
**Pull request by brianjohnson5972**
#5656 
Tool to take a valid block log directory and write the binary contents out as json.

### CODE REVIEW

File: plugins/blocklog_plugin/blocklog_plugin.cpp [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+blocklog_plugin::blocklog_plugin()
+:my(new blocklog_plugin_impl()) {
+}
+
+blocklog_plugin::~blocklog_plugin(){}
+
+void blocklog_plugin::set_program_options(options_description& cli, options_description& cfg)
+{
+   cfg.add_options()
+         ("blocks-dir,d", bpo::value<bfs::path>()->default_value("blocks"),
```
Comment (heifner) [inline plugins/blocklog_plugin/blocklog_plugin.cpp]: -d is not unique, collides with --data-dir

Comment (brianjohnson5972) [reply plugins/blocklog_plugin/blocklog_plugin.cpp]: will remove

File: plugins/blocklog_plugin/blocklog_plugin.cpp [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+
+   uint32_t block_num = (first_block < 1) ? 1 : first_block;
+   const uint32_t end_block = (last_block < end->block_num() ? last_block : end->block_num()) + 1;
+   signed_block_ptr next;
+   while((block_num < end_block) && (next = block_logger->read_block_by_num( block_num ))) {
+      fc::variant pretty_output;
+      abi_serializer::to_variant(*next,
+                                 pretty_output,
+                                 []( account_name n ) { return optional<abi_serializer>(); },
+                                 fc::microseconds(1000 * 1000));
```
Comment (heifner) [inline plugins/blocklog_plugin/blocklog_plugin.cpp]: use fc::microseconds::maximum()

Comment (heifner) [reply plugins/blocklog_plugin/blocklog_plugin.cpp]: Actually that doesn't work, so how about `fc::seconds(10)`

File: programs/eosio-blocklog/main.cpp [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+#include <eosio/chain/exceptions.hpp>
+
+#include <fc/log/logger_config.hpp>
+#include <fc/log/appender.hpp>
+#include <fc/exception/exception.hpp>
+#include <fc/filesystem.hpp>
+
+#include <boost/dll/runtime_symbol_info.hpp>
+#include <boost/exception/diagnostic_information.hpp>
+
```
Comment (heifner) [inline programs/eosio-blocklog/main.cpp]: Why make this a plugin? Can't you just move the logic to main and simply.

File: plugins/blocklog_plugin/blocklog_plugin.cpp (hunk 2 above)
Comment (heifner) [reply plugins/blocklog_plugin/blocklog_plugin.cpp]:
```
   signed_block_ptr next;
   fc::variant pretty_output;
   const fc::microseconds deadline = fc::seconds(10);
   while((block_num < end_block) && (next = block_logger->read_block_by_num( block_num ))) {
      abi_serializer::to_variant(*next,
                                 pretty_output,
                                 []( account_name n ) { return optional<abi_serializer>(); },
                                 deadline);
      const auto block_id = next->id();
      const uint32_t ref_block_prefix = block_id._hash[1];
      auto enhanced_object = fc::mutable_variant_object
                 ("block_num",next->block_num())
                 ("id", block_id)
                 ("ref_block_prefix", ref_block_prefix)
                 (pretty_output.get_object());
      fc::variant v( std::move(enhanced_object) );
      *out << fc::json::to_pretty_string(v) << "\n";
      ++block_num;
   }
```

Comment (heifner) [reply plugins/blocklog_plugin/blocklog_plugin.cpp]: Some preformance improvements. Also if you made pretty output optional you could write directly to output using `fc::to_stream( *out, v, fc::json::stringify_large_ints_and_doubles )`

File: plugins/blocklog_plugin/include/eosio/blocklog_plugin/blocklog_plugin.hpp [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+using chain::transaction;
+using chain::transaction_id_type;
+using fc::optional;
+using boost::container::flat_set;
+using chain::abi_def;
+using chain::abi_serializer;
+
+class blocklog_plugin : public plugin<blocklog_plugin> {
+public:
+   APPBASE_PLUGIN_REQUIRES()
```
[the remark this answers is not in the document]
Comment (brianjohnson5972) [reply plugins/blocklog_plugin/include/eosio/blocklog_plugin/blocklog_plugin.hpp]: this provides a method called plugin_requires that needs to be implemented

File: plugins/blocklog_plugin/blocklog_plugin.cpp [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+      output_blocks.open(output_file.generic_string().c_str());
+      if (output_blocks.fail()) {
+         std::ostringstream ss;
+         ss << "Unable to open file '" << output_file.string() << "'";
+         throw std::runtime_error(ss.str());
+      }
+      out = &output_blocks;
+   }
+   else
+      out = &std::cout;
```
Comment (heifner) [inline plugins/blocklog_plugin/blocklog_plugin.cpp]: Add `std::ios::sync_with_stdio(false);` for this case for a performance boost.

File: programs/eosio-blocklog/main.cpp [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+      const auto enhanced_object = fc::mutable_variant_object
+                 ("block_num",next->block_num())
+                 ("id", block_id)
+                 ("ref_block_prefix", ref_block_prefix)
+                 (pretty_output.get_object());
+      fc::variant v(std::move(enhanced_object));
+      if (no_pretty_print)
+         fc::json::to_stream(*out, v, fc::json::stringify_large_ints_and_doubles);
+      else
+         *out << fc::json::to_pretty_string(enhanced_object) << "\n";
```
Comment (heifner) [inline programs/eosio-blocklog/main.cpp]: `enhanced_object` -> `v`

File: programs/eosio-blocklog/main.cpp [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+
+      ++block_num;
+   }
+}
+
+void blocklog::set_program_options(options_description& cli)
+{
+   cli.add_options()
+         ("blocks-dir", bpo::value<bfs::path>()->default_value("blocks"),
+          "the location of the blocks directory (absolute path or relative to application data dir)")
```
Comment (heifner) [inline programs/eosio-blocklog/main.cpp]: This is no longer relative to application data dir.

File: programs/eosio-blocklog/main.cpp [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+         ("output-file,o", bpo::value<bfs::path>(),
+          "the file to write the block log output to (absolute or relative path)")
```
Comment (heifner) [inline programs/eosio-blocklog/main.cpp]: ".. If not specified output is to stdout."

File: programs/eosio-blocklog/main.cpp [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+         ("first", bpo::value<uint32_t>(&first_block)->default_value(1),
+          "the first block to log to file")
```
Comment (heifner) [inline programs/eosio-blocklog/main.cpp]: "the first block number to log to file"

File: programs/eosio-blocklog/main.cpp [hunk 10]
```diff
@@ -0,0 +1,158 @@
...
+         ("last", bpo::value<uint32_t>(&last_block)->default_value(std::numeric_limits<uint32_t>::max()),
+          "the last block (inclusive) to log to file")
```
Comment (heifner) [inline programs/eosio-blocklog/main.cpp]: "the last block number (inclusive) to log"

**Pull request merged**